### PR TITLE
Make fromTarPathToPosixPath fast

### PR DESCRIPTION
**Pull request by edmundnoble**
No reason to move to `String` so early when the `ByteString` ops are so much faster. In my application this brings this:

```
COST CENTRE            MODULE                  SRC                                          %time %alloc                                                                                                                                                                                                                                                                                                                                                
loadNewestFrom         Deps                    src/Deps.hs:73:1-55                           51.1   51.1                                                                                                                    
fromTarPathToPosixPath Codec.Archive.Tar.Types Codec/Archive/Tar/Types.hs:(321,1)-(329,36)   31.5   40.5                                                                                                                    unfoldEntries          Codec.Archive.Tar.Types Codec/Archive/Tar/Types.hs:(493,1)-(498,48)   11.4    4.5                                                                                                                    
hash                   Data.HashMap.Base       Data/HashMap/Base.hs:137:1-28                  2.5    3.6                                                                                                                    lookup                 Data.HashMap.Base       Data/HashMap/Base.hs:(434,1)-(448,29)          2.2    0.0     
```

To this:
```
COST CENTRE            MODULE                       SRC                                            %time %alloc                                                                                                             
                                                                                                                                                                                                                            
munch                  Distribution.Compat.ReadP    Distribution/Compat/ReadP.hs:(280,1)-(285,34)   28.6   29.0                                                                                                             
loadNewestFrom         Deps                         src/Deps.hs:73:1-55                             24.7   28.0                                                                                                             
unfoldEntries          Codec.Archive.Tar.Types      Codec/Archive/Tar/Types.hs:(492,1)-(497,48)     16.8    6.8                                                                                                             
sepBy1                 Distribution.Compat.ReadP    Distribution/Compat/ReadP.hs:360:1-45           12.8   11.5                                                                                                             
fromTarPathToPosixPath Codec.Archive.Tar.Types      Codec/Archive/Tar/Types.hs:(321,1)-(328,36)      2.7    8.7
```

### CODE REVIEW

**Review by emilypi — APPROVED** [review]

File: Codec/Archive/Tar/Types.hs [hunk 1]
```diff
@@ -343,7 +342,7 @@ fromTarPathToWindowsPath (TarPath namebs prefixbs) = adjustDirectory $
   where
     name   = BS.Char8.unpack namebs
     prefix = BS.Char8.unpack prefixbs
-    adjustDirectory | FilePath.Posix.hasTrailingPathSeparator name
+    adjustDirectory | BS.Char8.pack [FilePath.Posix.pathSeparator] `BS.isPrefixOf` namebs
```
Comment (Bodigrim) [inline Codec/Archive/Tar/Types.hs:345]: Err, why `pack [x]` and not `singleton x`?

File: Codec/Archive/Tar/Types.hs [hunk 2]
```diff
@@ -318,13 +318,12 @@ fromTarPath (TarPath namebs prefixbs) = adjustDirectory $
...
-                         ++ FilePath.Posix.splitDirectories name
+fromTarPathToPosixPath (TarPath namebs prefixbs) = adjustDirectory $ BS.Char8.unpack $
+   if BS.null prefixbs
+   then namebs
+   else prefixbs <> BS.Char8.pack [FilePath.Posix.pathSeparator] <> namebs
   where
-    name   = BS.Char8.unpack namebs
-    prefix = BS.Char8.unpack prefixbs
-    adjustDirectory | FilePath.Posix.hasTrailingPathSeparator name
+    adjustDirectory | BS.Char8.pack [FilePath.Posix.pathSeparator] `BS.isPrefixOf` namebs
```
Comment (Bodigrim) [inline Codec/Archive/Tar/Types.hs:326]: At the very least `singleton pathSeparator`, but I'd rather avoid `isPrefixOf`, when a simpler function would suffice.

File: Codec/Archive/Tar/Types.hs [hunk 3]
```diff
@@ -318,13 +318,12 @@ fromTarPath (TarPath namebs prefixbs) = adjustDirectory $
 -- operating system, eg to perform portability checks.
 --
 fromTarPathToPosixPath :: TarPath -> FilePath
-fromTarPathToPosixPath (TarPath namebs prefixbs) = adjustDirectory $
-  FilePath.Posix.joinPath $ FilePath.Posix.splitDirectories prefix
-                         ++ FilePath.Posix.splitDirectories name
+fromTarPathToPosixPath (TarPath namebs prefixbs) = adjustDirectory $ BS.Char8.unpack $
+   if BS.null prefixbs
+   then namebs
+   else prefixbs <> BS.Char8.pack [FilePath.Posix.pathSeparator] <> namebs
```
Comment (Bodigrim) [inline Codec/Archive/Tar/Types.hs:324]: AFAICT this shortcut relies on the fact that `namebs` and `prefixbs` are stored in POSIX format. Do we have tests for this function? Did anyone test it on Windows?

File: Codec/Archive/Tar/Types.hs (hunk 2 above)
Comment (Bodigrim) [reply Codec/Archive/Tar/Types.hs:326]: Sorry, I don't understand this. The code used to check that `name` has a trailing `/`, but now we check that `namebs` has a leading `/`?.. It does not look right to me at all.

File: Codec/Archive/Tar/Types.hs (hunk 3 above)
Comment (edmundnoble) [reply Codec/Archive/Tar/Types.hs:324]: I believe that is not new to this code. The old code used `FilePath.Posix.splitDirectories`.

File: Codec/Archive/Tar/Types.hs (hunk 2 above)
Comment (edmundnoble) [reply Codec/Archive/Tar/Types.hs:326]: 🤦

**Review by edmundnoble — COMMENTED** [review]
Thank you for your review @Bodigrim. I hope I have addressed your concerns with my responses and my ~~forthcoming~~ PR, #60.

File: Codec/Archive/Tar/Types.hs (hunk 2 above)
Comment (edmundnoble) [reply Codec/Archive/Tar/Types.hs:326]: What function did you have in mind?

**Pull request merged**